### ISSUE TITLE
Atlas has kind of a high security risk/attack surface

### ISSUE THREAD

**Issue by 205g0** [ariga/atlas#1743]
Just realized that you even can add the production url (not just dev) to atlas.hcl, e.g., in `env "prod"`. If some dev on your team commits an atlas.hcl incl. the production url to git/Github this would be really unfortunate. I mean the connection string to a production db might be the most important thing in an entire company.

In contrast, you would want to commit atlas.hcl to git and the project's repo, nobody wants to reenter all the config data. So now and even worse, devs tend to just commit atals.hcl (because there's also no warning in your docs and it must be meant like this).

Hope, I got something wrong.

Related to https://github.com/ariga/atlas/issues/1742

**Comment by a8m** [comment]
Hey, @205g0 👋 

Thanks for opening this issue, but unfortunately, you got missed some major parts of the doc. 


The `atlas.hcl` configuration file allows you to define one or more environment blocks along with additional global configurations. Although it's possible to use the database URL as a raw string (like in any other program), the recommended approach is to either pass it as a CLI argument, project variable, environment variable, or let Atlas read it from secret storage such as AWS/GCP SecretManager. All of these options are described in detail on our documentation website: https://atlasgo.io/atlas-schema/projects.

Closing, but please feel free to reopen or join our [Discord server](https://discord.gg/zZ6sWVg6NT) if you need more help with this.

**Comment by 205g0** [comment]
tldr:: I'd ask to reopen, I think Atlas has still security and GDPR issues. Thanks. Please read all for a better understanding.

>  it's possible to use the database URL as a raw string (like in any other program)

It shouldn't be possible in the first place because atlas.hcl is not meant as a .gitignore file becauee it contains too much config. Also not for dev urls (they carry GDPR risks, see below)

> the recommended approach is to either pass it as a CLI argument

Then, it's in Bash's history, not much better (a bit)

> or let Atlas read it from secret storage such as AWS/GCP SecretManager

- Not everyone is on AWS/GCP but this is not the actual problem, read on
- With others it doesn't work, e.g. you can't pull env vars into a hcl file from Cloudflare Secrets Store; latter also offers only namespaces for prod and preview but not local
- The real issue: .env meant for local env vars; and for many workflows you push db changes from a local dev env to a dev db in the cloud, e.g. Planretscale; then, from dev to production isn't handled by Atlas anymore but by Planetscale and it's own schema migration; but long story short, in your local dev env where prod is simulated (e.g. in wrangler or in Docker) you don't access remote secrets stores because there are typically not for local dev envs
- So, your solution covers just a very specific workflow but nothing else, e.g. Cloudflare, Vercel, Neon, et.c
- The other options http (even less secure), file (more cumbersome and another file I have to add to .gitignore) are unfortunately not better
- Using input vars with atlas.hcl and the CLI face the aforementioned problem with the Bash history

Again, the very immediate risk (security and GDPR) of all Atlas UX limitations is still there: Devs disclose the dev db url to git/GitHub, they might think:

- "dev db urls of db branches are ephemeral, so what"
- "nobody cares, if they are on git"
- "it's just a branch"
- "fuzzing around with AWS/GCP secrets manager is too much hassle for an ephemeral dev string haha" or "who is using remote secrets store for dev db strings" or "who is still an expensive AWS/GCP when Cloudflare gives you all for a fraction?"

So, the dev tends to commit the dev string to git and calls it a day. Then, an attacker can easily get all full production data and the company needs to file a GDPR breach. 

> Closing, but please feel free to reopen or join our [Discord server](https://discord.gg/zZ6sWVg6NT) if you need more help with this.

No offense but if you were interested in a convo you hadn't instantly closed and renamed the issue. So, if you ask, yes, I'd like to reopen. However, I do not expect any further conversation because of your initial signaling.

**Comment by a8m** [comment]
Hey, @205g0,
I'll recommend checking our deployment guides: https://atlasgo.io/guides/deploying/intro. We try to cover most of the common use cases. In all examples, secrets are read/passed using standard tools such as Vault, SecretManager, Kubernetes secrets (files or env), etc. If you'll share more about your specific use case, I can try to help with this. 

> It shouldn't be possible in the first place because atlas.hcl is not meant as a .gitignore file becauee it contains too much config. Also not for dev urls (they carry GDPR risks, see below)

I beg to differ, as we can tell this on almost any other tool, such as TF or Pulumi, or GitHub Actions. We expect users to use the recommended ways to inject secrets into Atlas.

> No offense but if you were interested in a convo you hadn't instantly closed and renamed the issue. So, if you ask, yes, I'd like to reopen. However, I do not expect any further conversation because of your initial signaling.

I didn't mean to be rude. I close issues when they are resolved from my side and always invite users to Discord or to reopen if they want to continue the discussion. Handling a few issues (and sometimes more than 10) every day, I find it easier to handle OSS engagement this way, something I do for years.

TBH, I didn't really understand your concerns or suggestions. What do you suggest and what are the limitations you're facing atm?

**Comment by 205g0** [comment]
Thanks for the prompt reply and apologies for my harsh tone, I need to work on that.

Just an idea, what about letting the CLI accepting input vars for atlas.hcl via `--var` not just from a file but from .env? Would this be a big effort to implement?

**Comment by 205g0** [comment]
`.env` is just a file full of env vars, e.g.:
```
US_EAST_1=mysql://user:pass...
US_WEST_2=mysql://user:pass..
....
```

**Comment by rotemtam** [comment]
Hey @205g0 

Thanks for reaching out. @a8m and I take security and privacy concerns very seriously so this topic is important to us. 

I'm not sure, however, how this issue pertains to GDPR which is European Privacy Regulation.

We have added support for numerous ways to supply credentials from industry-standard ways of providing secrets, so I'm sure you can understand that the way you phrased your criticism doesn't exactly invite discussion. 

Putting all of this aside, addressing `.env` file concretely, I know it's a popular format in the Node.s community, and used by some other tools as well. We can consider adding support for this in Atlas as well but prefer not to tie ourselves to the ecosystem of a specific programming language runtime. 

Having said that, to unblock you, here's how you can use `.env` files today:

Suppose the `.env` file is adjacent to your `atlas.hcl` file. 

```env
URL=mysql://root:pass@localhost:3306/myapp
```

You `atlas.hcl` file: 

```hcl
variable "envfile_path" {
  type = string
}

data "runtimevar" "env" {
  url = "file://${var.envfile_path}"
}

locals {
  envlines = split("\n", data.runtimevar.env)
  envfile = { for line in local.envlines : split("=", line)[0] => split("=", line)[1] }
}

env {
  name = atlas.env
  url = local.envfile["URL"]
}
```

Now, from your local workstation use: 
```shell
atlas schema inspect --env x --var envfile_path="$(pwd)/.env"
```

It's not exactly what you were hoping for, but it will achieve your goals, I think.

Let me know how this looks.

**Comment by 205g0** [comment]
Thanks for the help! However, it doesn't work. I have a proper connection url which works perfectly with the `-u` flag and ends with `...?tls=true` but I get...
```
Error: mysql: query system variables: Error 1105 (HY000): unknown error: Code: UNAVAILABLE
server does not allow insecure connections, client must use SSL/TLS
```
Tinkered a bit with the HCL file but couldn't get it to work, ideas what I did wrong?

**Comment by a8m** [comment]
It happens because @rotemtam’s example uses split on `=`. I will share a updated example later today.

**Comment by 205g0** [comment]
.Ah right, thanks for the hint, I've fixed it, the `envline` file needs to be replaced by...
```
  envfile = { for line in local.envlines : split("=", line)[0] => regex("=(.*)", line)[0] }
```
Wonderful!

**Comment by 205g0** [comment]
@a8m: Zuckerberg once called Twitter ‘a clown car that fell into a gold mine’.

Not that Atlas is the clown car—no, it's currently the best decl. migration tool out there--but somehow I've to think of Zuck's quote when using Atlas.

- Declarative migrations are the future and because the majority doesn't get it and still work with lol ancient migration tools, the market needs to be educated (entry barrier 1)
- Then, HCL, while powerful, it's popular just in few niches, even in the dominant k8s world, HCL is not a single thing; while it's amazing to write scripts in your config, it's also a bit odd (syntax, debugging, testing, etc..); so here, we have entry barrier 2
- Do you know how big the TypeScript on the backend/node market is? Did you check Next.js recent download figures? And it's just one of many node backend frameworks (or SSR frameworks), you will MISS all these potential customers because not supporting something simple and learned like `.env` which is a no-brainer;  they rather do old-school migrations instead of going to the hassle of setting up Atlas + after reading the Go fanboism on the Github README; there is even the Go mascot, omg, so if you want to be lang agnostic, remove all this Go stuff and treat all langs equal; also the usage of Go as a glue layer between your db and your app or web server is a fraction of node/TypeScript

@rotemtam re GDPR:
Letting one putting raw secrets in a config file (atlas.hcl) is an antipattern. You can debate with me, I don't care but if I were you I wouldn't want to discuss this with legal authorities when the first breach is out and the breach gets attached to Atlas.hcl which lets you put raw DB strings (seriously?) in a config file. DB strings are holy and when reading the docs or using the Atlas CLI or atlas.hcl, it doesn't feel like you guys are aware what relevance the DB connections string actually have

Also, you position Atlas kind of enterprise-y (docs but especially the pricing with the contact us button), so then you need to do your homework re GDPR. But that's your call, I'm totally fine with this HCL script, I'll never touch that again anyway, haha.

Please take my feedback not as a rant. I am huge fan of you guys but you miss your potential. I've tested all push tools in and out and you are far ahead

**Comment by a8m** [comment]
@205g0, I feel like the conversation got a bit too emotional, and honestly, I'm not entirely sure why. You can join our [Discord server](https://discord.gg/zZ6sWVg6NT) and we can continue discussing there, hopefully in a more friendly manner 😃 

Just to highlight, your suggestion wasn't rejected. As Rotem pointed out earlier, we will consider your suggestion and explore deployment systems that promote this approach (I'll give a look at the examples you mentioned - Cloudflare, Vercel, Neon). The reason we haven't supported it until now is that we haven't received such a request from any company or user we work with.

**Comment by 205g0** [comment]
>  feel like the conversation got a bit too emotional,

Thanks for your reply, I'm sorry that you feel that way, I tried to give constructive feedback. Doing that right isn't easy but I had/have good intentions (I just recommended Atlas to two friends haha). Maybe it's your perception or you who gets emotional about my words. I just said, in a too long post, that Atlas is underrated and you should fix that.

> we haven't received such a request from any company or user we work with.

C'mon, you know why:

- Entry barrier 1, there not many using decl. migrations yet. Just read the comments of your last HN submission, the top commentor talked the product down because he didn't get it at all, it was so sad to read his words misleading thousands of users, even worse most upvoted this nonsense
- Because only Go users your product and the reason is in the next bullet
- This Go obsession (this is not emotional but important to make my point clear}: When a TypeScript user sees the Go mascot they instant press command-w or go back; they do not realize, oh I can use this too? No, they think first, it's just for Go users and TS users have a deep aversion against Go; this is not my opinion but if you know this ecosystem they make fun of Go 24/7 ("type system is a joke to TS'", generics were just recently introduced, weird syntax/ugliness, etc. again not my opinion, I find Go great, just not using it for CRUD but for k8s); you need to understand how the biggest developer group sees you product or better, don't see it because it's not on their radar at all; do you want to boost adoption of Atlas or be Go evangelist? Make a choice
- You create this cognitive dissonance with other dev groups as well, even more with Java, RoR and Rust

> You can join our [Discord server](https://discord.gg/zZ6sWVg6NT) and we can continue discussing there, hopefully in a more friendly manner

Sure anytime, apologies for being so direct but I'm actually to enthusiastic because Atlas is just a great product and it took me months to find it (now ask yourself why). Yeah, you missed .env but other than that you do a great job. 🙂
